core/string/ustring.cpp

```diff
@@ -5365,64 +5409,16 @@ bool String::is_relative_path() const {
 }
 
 String String::get_base_dir() const {
-	int end = 0;
+	const int drive_end_idx = find_drive_end_idx(*this);
+	const int last_separator_idx = MAX(rfind_char('/'), rfind_char('\\'));
 
-	// URL scheme style base.
-	int basepos = find("://");
-	if (basepos != -1) {
-		end = basepos + 3;
-	}
-
-	// Windows top level directory base.
-	if (end == 0) {
-		basepos = find(":/");
-		if (basepos == -1) {
-			basepos = find(":\\");
-		}
-		if (basepos != -1) {
-			end = basepos + 2;
-		}
-	}
-
-	// Windows UNC network share path.
-	if (end == 0) {
-		if (is_network_share_path()) {
-			basepos = find_char('/', 2);
-			if (basepos == -1) {
-				basepos = find_char('\\', 2);
-			}
-			int servpos = find_char('/', basepos + 1);
-			if (servpos == -1) {
-				servpos = find_char('\\', basepos + 1);
-			}
-			if (servpos != -1) {
-				end = servpos + 1;
-			}
-		}
-	}
-
-	// Unix root directory base.
-	if (end == 0) {
-		if (begins_with("/")) {
-			end = 1;
-		}
-	}
-
-	String rs;
-	String base;
-	if (end != 0) {
-		rs = substr(end, length());
-		base = substr(0, end);
-	} else {
-		rs = *this;
-	}
-
-	int sep = MAX(rs.rfind_char('/'), rs.rfind_char('\\'));
-	if (sep == -1) {
-		return base;
+	if (last_separator_idx < drive_end_idx) {
```

Comment:
the if + 2 returns can be squashed into : `return substr(0, MAX(last_separator_idx, drive_end_idx))`

Reply:
Oh, you're right, that works because `substr` already does non-copy logic. Makes sense to me!